sqlglot/helper.py

```diff
@@ -1,48 +1,112 @@
+from __future__ import annotations
+
 import inspect
 import logging
 import re
 import sys
 import typing as t
+from collections.abc import Collection
 from contextlib import contextmanager
 from copy import copy
 from enum import Enum
 
+if t.TYPE_CHECKING:
+    from sqlglot.expressions import Expression, Table
+
+    T = t.TypeVar("T")
+    E = t.TypeVar("E", bound=Expression)
+
 CAMEL_CASE_PATTERN = re.compile("(?<!^)(?=[A-Z])")
 logger = logging.getLogger("sqlglot")
 
 
 class AutoName(Enum):
-    def _generate_next_value_(name, _start, _count, _last_values):
+    """This is used for creating enum classes where `auto()` is the string form of the corresponding value's name."""
+
+    def _generate_next_value_(name, _start, _count, _last_values):  # type: ignore
         return name
 
 
-def list_get(arr, index):
+def sequence_get(seq: t.Sequence[T], index: int) -> t.Optional[T]:
+    """Returns the value in `seq` at position `index`, or `None` if `index` is out of bounds."""
     try:
-        return arr[index]
+        return seq[index]
     except IndexError:
         return None
 
 
-def ensure_list(value):
+def ensure_list(value: t.Tuple[T] | t.List[T] | T) -> t.List[T]:
+    """
+    Ensures that a value is a list, otherwise casts or wraps it into one.
+
+    Args:
+        value: the value of interest.
+
+    Returns:
+        The value cast as a list if it's a list or a tuple, or else the value wrapped in a list.
+    """
     if value is None:
         return []
-    return value if isinstance(value, (list, tuple, set)) else [value]
+    elif isinstance(value, (list, tuple)):
+        return list(value)
 
+    return [value]
+
+
+@t.overload
+def ensure_collection(value: t.Collection[T]) -> t.Collection[T]:
+    ...
+
+
+@t.overload
+def ensure_collection(value: T) -> t.Collection[T]:
+    ...
+
+
+def ensure_collection(value):
```

Comment:
can't this just be 

```
def ensure_collection(value: T | t.Collection[T]) -> t.Collection[T] 
```

or does the type checker get confused?

Reply:
> or does the type checker get confused?

That seems to be the case:
```python
sqlglot/helper.py:88: error: List item 0 has incompatible type "Union[T, str, bytes]"; expected "T"
```

Reply:
If value is `T | t.Collection[T]`, `T` will swallow up any type.
```python
ensure_collection([])  # the return type would be t.Collection[list]
```

Reply:
Right, whereas using overloads we expect mypy to explore all variants in the same order as they're declared, correct?

Reply:
I think so. Might want to verify that.

Reply:
Check, see here: https://mypy.readthedocs.io/en/stable/more_types.html#type-checking-calls-to-overloads

TL;DR
```
When you call an overloaded function, mypy will infer the correct return type by picking the best matching variant, after taking into consideration both the argument types and arity. However, a call is never type checked against the implementation. This is why mypy will report calls like mouse_event(5, 25, 3) as being invalid even though it matches the implementation signature.

If there are multiple equally good matching variants, mypy will select the variant that was defined first. [...]
```